### PR TITLE
Subscription graphiql support

### PR DESCRIPTION
**Pull request by junminstorage**
See https://github.com/graphql/express-graphql/issues/390#issuecomment-371269023 for usage

### CODE REVIEW

File: src/renderGraphiQL.js [hunk 1]
```diff
@@ -72,6 +82,7 @@ add "&raw" to the end of the URL within a browser.
   <script src="//cdn.jsdelivr.net/react/15.4.2/react.min.js"></script>
   <script src="//cdn.jsdelivr.net/react/15.4.2/react-dom.min.js"></script>
   <script src="//cdn.jsdelivr.net/npm/graphiql@${GRAPHIQL_VERSION}/graphiql.min.js"></script>
```
Comment (junminstorage) [inline src/renderGraphiQL.js]: ideally we should manage the react and other client-side libs via npm or bower, npm is recommended, then browserify and create bundle,  in that way, some issue like #112 can be resolved.

**Comment by IvanGoncharov** [comment]
@junminstorage Thanks for PR 👍 
As I understand goal of this PR is to add subscription support to the GraphiQL.

I think it's pretty big step to add `subscriptions-transport-ws` as a dependency since it's Apollo-specific and also marked as `Work in progress!` in the README.

Anyway, I think making such big decigions is beyound my current role in this project.

@leebyron What do you think about this feature?

**Comment by calvernaz** [comment]
Is anything we can do to move this one forward?

**Comment by junminstorage** [comment]
'subscriptions-transport-ws' is generic graphql subscription spec implementation, I am sure why we need to reinvent a wheel. 

and this PR answered a well-liked community's request per original question from https://github.com/graphql/express-graphql/issues/390#issue-253919382

**Comment by tylerreckart** [comment]
It would be nice to see some movement on this PR

**Comment by junminstorage** [comment]
ping @IvanGoncharov and @leebyron here. let us know what we need to do here to get it resolved.

**Comment by vrochamaia** [comment]
Any updates on this?

**Comment by linxiaowu66** [comment]
Any updates on this?

**Comment by benjie** [comment]
Thank you for your pull request to GraphiQL, and your interest in the project. GraphiQL does not concern itself with the transport layer: HTTP, websockets, IPC, web workers, etc — GraphiQL is used in many different contexts, and it's not always HTTP as the transport so we do not tend to merge HTTP-specific features. Instead we abstract details of the transport away, via the `fetcher` function.

GraphiQL already supports subscriptions via a subscriptions-aware `fetcher`. A subscriptions-aware fetcher might look something like this:

```js
let activeSubscription;

/*
 * Future queries, mutations and subscriptions should cancel previous
 * queries, mutations and subscriptions
 */
const cancelSubscription = () => {        
  if (activeSubscription) {
    activeSubscription.unsubscribe();
    activeSubscription = null;
  }
};

/*
 * Our subscription aware fetcher determines if the request is a subscription
 * or not. If it is, it uses the `subscriptionsClient`, otherwise it falls
 * back to the regular fetcher.
 */
const subscriptionAwareFetcher = params => {
  cancelSubscription();
  if (isSubscription(params) && subscriptionsClient) {
    return {
      subscribe: observer => {
        cancelSubscription();
        activeSubscription = subscriptionsClient.request(params).subscribe(observer);
        return activeSubscription;
      }
    };
  } else {
    /* Do whatever your old fetcher did: */
    return regularFetcher(params);
  }
}
```

[Here is the fetcher I use in PostGraphile](https://github.com/graphile/postgraphile/blob/fe6e91a7ffea66468a5747f440322bc5f8506fc1/postgraphiql/src/components/PostGraphiQL.js#L255-L289), for example. I would not recommend copying it verbatim ;)

We'd definitely appreciate pull requests to the documentation to make this clearer (preferably with a working example). There are many places where GraphiQL's subscriptions support could be improved (for example it currently outputs `[object Object]` rather than the actual error when subscription errors occur), PRs for those would be welcome also.

I'm going to close this, as the functionality is already available in GraphiQL.

**Comment by benjie** [comment]
And now I'm going to re-open it, blushing heavily, because I just realised this is the `express-graphql` repo, not the `graphiql` repo. 😳 

Gosh, is my face red right now.

**Comment by nileio** [comment]
thanks for pushing this through. Anyone from the team could provide us with an update on this please? keen to get subscriptions to work

**Comment by junminstorage** [comment]
Hi All, I am original author for this PR. per the https://github.com/graphql/express-graphql/pull/436#issuecomment-388960292, the roadblocker is the dependency on subscriptions-transport-ws due to its repo was not under MIT license? Now this repo is under MIT now. Let me know if anything I can do to move it forward.

**Comment by acao** [comment]
that would be fantastic if you could do that @junminstorage !  going to add myself and daniel as reviewers so we can get this through. apologies the GraphiQL docs are so sparse on subscriptions, we need to add that to the docs/examples very badly.

**Comment by junminstorage** [comment]
Awesome I am going to working on this PR this week. expect to have it ready by early next week.

**Comment by cannc4** [comment]
Looking forward to it!

**Comment by junminstorage** [comment]
closed it and replaced by https://github.com/graphql/express-graphql/pull/686

**Comment by junminstorage** [comment]
replaced by #687

**Pull request closed without merging**